### PR TITLE
feat: contract manager resubmit #ntrn-84

### PR DESCRIPTION
**Pull request by NeverHappened**
https://hadronlabs.atlassian.net/browse/NTRN-84

### CODE REVIEW

**Comment by NeverHappened** [comment]
https://github.com/neutron-org/neutron-tests/actions/runs/5896212533

**Comment by NeverHappened** [comment]
https://github.com/neutron-org/neutron-tests/actions/runs/5896234877

File: x/contractmanager/keeper/failure.go [hunk 1]
```diff
@@ -54,3 +55,60 @@ func (k Keeper) GetAllFailures(ctx sdk.Context) (list []types.Failure) {
...
+}
+
+// ResubmitFailure tries to call sudo handler for contract with same parameters as initially.
+func (k Keeper) ResubmitFailure(ctx sdk.Context, contractAddr sdk.AccAddress, failure *types.Failure) error {
+	if failure.Packet == nil {
+		return errors.Wrapf(types.IncorrectFailureToResubmit, "cannot resubmit failure without packet info; failureId = %d", failure.Id)
+	}
+
+	switch failure.GetAckType() {
+	case "ack":
```
Comment (mmikev) [inline x/contractmanager/keeper/failure.go]: Let's make it a constant

File: x/contractmanager/keeper/failure.go [hunk 2]
```diff
@@ -54,3 +55,60 @@ func (k Keeper) GetAllFailures(ctx sdk.Context) (list []types.Failure) {
...
+			if err != nil {
+				return errors.Wrapf(types.FailedToResubmitFailure, "cannot resubmit failure ack response; failureId = %d; err = %s", failure.Id, err)
+			}
+		} else {
+			_, err := k.SudoError(ctx, contractAddr, *failure.Packet, failure.Ack.GetError())
+			if err != nil {
+				return errors.Wrapf(types.FailedToResubmitFailure, "cannot resubmit failure ack error; failureId = %d; err = %s", failure.Id, err)
+			}
+		}
+	case "timeout":
```
Comment (mmikev) [inline x/contractmanager/keeper/failure.go]: Let's make it a constant

File: x/contractmanager/keeper/failure.go [hunk 3]
```diff
@@ -54,3 +55,60 @@ func (k Keeper) GetAllFailures(ctx sdk.Context) (list []types.Failure) {
...
+	}
+
+	switch failure.GetAckType() {
+	case "ack":
+		if failure.GetAck() == nil {
+			return errors.Wrapf(types.IncorrectFailureToResubmit, "cannot resubmit failure without acknowledgement; failureId = %d", failure.Id)
+		}
+		if failure.GetAck().GetError() == "" {
+			_, err := k.SudoResponse(ctx, contractAddr, *failure.Packet, failure.Ack.GetResult())
+			if err != nil {
```
Comment (mmikev) [inline x/contractmanager/keeper/failure.go]: Combine it into one line please

```
if _, err := k.SudoResponse(ctx, contractAddr, *failure.Packet, failure.Ack.GetResult()); err != nil {
```

File: x/contractmanager/keeper/failure.go [hunk 4]
```diff
@@ -54,3 +55,60 @@ func (k Keeper) GetAllFailures(ctx sdk.Context) (list []types.Failure) {
...
+				return errors.Wrapf(types.FailedToResubmitFailure, "cannot resubmit failure ack response; failureId = %d; err = %s", failure.Id, err)
+			}
+		} else {
+			_, err := k.SudoError(ctx, contractAddr, *failure.Packet, failure.Ack.GetError())
+			if err != nil {
```
Comment (mmikev) [inline x/contractmanager/keeper/failure.go]: Combine it into one line please

```
if _, err := k.SudoError(ctx, contractAddr, *failure.Packet, failure.Ack.GetError()); err != nil {
```

File: x/contractmanager/keeper/failure.go [hunk 5]
```diff
@@ -54,3 +55,60 @@ func (k Keeper) GetAllFailures(ctx sdk.Context) (list []types.Failure) {
...
+				return errors.Wrapf(types.FailedToResubmitFailure, "cannot resubmit failure ack error; failureId = %d; err = %s", failure.Id, err)
+			}
+		}
+	case "timeout":
+		_, err := k.SudoTimeout(ctx, contractAddr, *failure.Packet)
+		if err != nil {
```
Comment (mmikev) [inline x/contractmanager/keeper/failure.go]: Combine it into one line please

```
if _, err := k.SudoTimeout(ctx, contractAddr, *failure.Packet); err != nil {
```

File: x/transfer/ibc_handlers.go [hunk 6]
```diff
@@ -107,7 +108,7 @@ func (im IBCModule) HandleAcknowledgement(ctx sdk.Context, packet channeltypes.P
 	}
 
 	if err != nil {
-		im.ContractManagerKeeper.AddContractFailure(ctx, packet.SourceChannel, senderAddress.String(), packet.GetSequence(), "ack")
+		im.ContractManagerKeeper.AddContractFailure(ctx, &packet, senderAddress.String(), "ack", &ack)
```
Comment (mmikev) [inline x/transfer/ibc_handlers.go]: Use the constant for "ack" i told above please

File: x/transfer/ibc_handlers.go [hunk 7]
```diff
@@ -140,11 +141,11 @@ func (im IBCModule) HandleTimeout(ctx sdk.Context, packet channeltypes.Packet, r
 	}
 
 	cacheCtx, writeFn, newGasMeter := im.createCachedContext(ctx)
-	defer im.outOfGasRecovery(ctx, newGasMeter, senderAddress, packet, data, "timeout")
+	defer im.outOfGasRecovery(ctx, newGasMeter, senderAddress, packet, data, "timeout", nil)
```
Comment (mmikev) [inline x/transfer/ibc_handlers.go]: Use the constant for "timeout" i told above please

**Review by pr0n00gler — CHANGES_REQUESTED** [review]

**Comment by NeverHappened** [comment]
TODO_RELAYER_NODE

**Comment by NeverHappened** [comment]
https://github.com/neutron-org/neutron-tests/actions/runs/5929600149

**Comment by NeverHappened** [comment]
https://github.com/neutron-org/neutron-tests/actions/runs/5936584543

**Comment by NeverHappened** [comment]
https://github.com/neutron-org/neutron-tests/actions/runs/5939048525

**Review by pr0n00gler — APPROVED** [review]

**Comment by NeverHappened** [comment]
https://github.com/neutron-org/neutron-tests/actions/runs/5939743946

File: wasmbinding/bindings/query.go [hunk 8]
```diff
@@ -108,6 +113,10 @@ type QueryMinIbcFeeResponse struct {
 	MinFee feerefundertypes.Fee `json:"min_fee"`
 }
 
+type FailuresResponse struct {
```
Comment (swelf19) [inline wasmbinding/bindings/query.go]: lets move request and response next each other

File: wasmbinding/message_plugin.go [hunk 9]
```diff
@@ -863,6 +871,39 @@ func (m *CustomMessenger) removeSchedule(ctx sdk.Context, contractAddr sdk.AccAd
...
+	if err != nil {
+		ctx.Logger().Error("failed to resubmitFailure",
+			"from_address", contractAddr.String(),
+			"error", err,
+		)
+		return nil, nil, errors.Wrap(err, "failed to resubmitFailure")
+	}
+
+	resp := bindings.ResubmitFailureResponse{}
+	data, err := json.Marshal(&resp)
```
Comment (swelf19) [inline wasmbinding/message_plugin.go:891]: we marshal a lot of empty responses here. Do we realy need process every error? or we can save a couple of lines with 100% working json.MustMarshal? Or even return `[]byte("{}")`

File: wasmbinding/message_plugin.go [hunk 10]
```diff
@@ -863,6 +871,39 @@ func (m *CustomMessenger) removeSchedule(ctx sdk.Context, contractAddr sdk.AccAd
 	return nil, [][]byte{data}, nil
 }
 
+func (m *CustomMessenger) resubmitFailure(ctx sdk.Context, contractAddr sdk.AccAddress, resubmitFailure *bindings.ResubmitFailure) ([]sdk.Event, [][]byte, error) {
+	failure, err := m.ContractmanagerKeeper.GetFailure(ctx, contractAddr, resubmitFailure.FailureId)
+	if err != nil {
+		return nil, nil, errors.Wrap(sdkerrors.ErrNotFound, "no failure found to resubmit")
+	}
+
+	if failure.Address != contractAddr.String() {
```
Comment (swelf19) [inline wasmbinding/message_plugin.go]: looks like unnecessary check, we `GetFailure` for a certain address. How is it possible we get something else.

File: x/contractmanager/keeper/failure.go [hunk 11]
```diff
@@ -54,3 +54,57 @@ func (k Keeper) GetAllFailures(ctx sdk.Context) (list []types.Failure) {
...
+	}
+	var res types.Failure
+	k.cdc.MustUnmarshal(bz, &res)
+
+	return &res, nil
+}
+
+// ResubmitFailure tries to call sudo handler for contract with same parameters as initially.
+func (k Keeper) ResubmitFailure(ctx sdk.Context, contractAddr sdk.AccAddress, failure *types.Failure) error {
+	if failure.Packet == nil {
```
Comment (swelf19) [inline x/contractmanager/keeper/failure.go:74]: During the migration we set `failure.Packet = nil` for old failures, is there a way to process those failures and remove them from the storage

File: wasmbinding/test/custom_message_test.go [hunk 12]
```diff
@@ -593,6 +598,101 @@ func (suite *CustomMessengerTestSuite) TestAddRemoveSchedule() {
...
+	suite.contractAddress = suite.InstantiateReflectContract(suite.ctx, suite.contractOwner, codeID)
+	suite.Require().NotEmpty(suite.contractAddress)
+
+	// Add failure
+	packet := ibcchanneltypes.Packet{}
+	ack := ibcchanneltypes.Acknowledgement{
+		Response: &ibcchanneltypes.Acknowledgement_Error{Error: "Error"},
+	}
+	failureID := suite.messenger.ContractmanagerKeeper.GetNextFailureIDKey(suite.ctx, testutil.TestOwnerAddress)
+	suite.messenger.ContractmanagerKeeper.AddContractFailure(suite.ctx, &packet, testutil.TestOwnerAddress, "timeout", &ack)
```
Comment (swelf19) [inline wasmbinding/test/custom_message_test.go]: Actually there is no `ack` on timeout, and pls replace "timeout" with the const

File: wasmbinding/message_plugin.go (hunk 9 above)
Comment (mmikev) [reply wasmbinding/message_plugin.go:891]: Why is it empty by the way? Shouldn't we put a failure id in the response?

Comment (swelf19) [reply wasmbinding/message_plugin.go:891]: i think its not necessary, a user sends a failureID to resubmit

Comment (mmikev) [reply wasmbinding/message_plugin.go:891]: But he may want to do something after the resubmit based on that id. I actually think we should always return id of an entity in our bindings

**Comment by NeverHappened** [comment]
https://github.com/neutron-org/neutron-tests/actions/runs/5998807948

**Review by swelf19 — APPROVED** [review]

**Review by pr0n00gler — APPROVED** [review]

**Review by quasisamurai — APPROVED** [review]

**Pull request merged**